lib/Model/fn.js

```diff
@@ -109,12 +108,24 @@ Fns.prototype.start = function(name, path, inputPaths, fns, options) {
   fns || (fns = this.nameMap[name] || defaultFns[name]);
   var fn = new Fn(this.model, name, path, inputPaths, fns, options);
   this.fromMap[path] = fn;
+  for (var i = 0; i < fn.inputsSegments.length; i++) {
+    var inputSegements = fn.inputsSegments[i];
+    this.inputListeners.addListener(inputSegements, fn);
+  }
+  this.outputListeners.addListener(fn.fromSegments, fn);
   return fn.onInput();
 };
 
 Fns.prototype.stop = function(path) {
```

Comment:
When Derby destroys a component, it calls `model.destroy('$components.___')`, which calls `Model#_stopAll`, which loops over `fromMap` and calls `Fns#stop` on each function at and underneath that point.

https://github.com/derbyjs/racer/blob/5e6ac76528097b55b40d6edbe442ef917f0f48fd/lib/Model/fn.js#L85-L93

A nice additional optimization would be to take advantage of the new tree structure for that, too. I think we'd still need to walk the `outputListeners` subtree, to clean up the input listener tree of all output listeners that are getting removed.